services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/ReceiveQueueBuffer.java

```diff
@@ -0,0 +1,192 @@
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.Message;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
+
+@SdkInternalApi
+public class ReceiveQueueBuffer {
+
+    private final ScheduledExecutorService executor;
+    private final SqsAsyncClient sqsClient;
+    private final ResponseBatchConfiguration config;
+    private final String queueUrl;
+    private final  QueueAttributesManager queueAttributesManager;
+
+    private final Queue<AsyncReceiveMessageBatch> finishedTasks = new ConcurrentLinkedQueue<>();
+    private final Queue<ReceiveMessageCompletableFuture> futures = new ConcurrentLinkedQueue<>();
+
+    private final AtomicInteger inflightReceiveMessageBatches = new AtomicInteger(0);
+    private final AtomicBoolean shutDown = new AtomicBoolean(false);
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    public ReceiveQueueBuffer(ScheduledExecutorService executor, SqsAsyncClient sqsClient,
+                              ResponseBatchConfiguration config, String queueUrl, QueueAttributesManager queueAttributesManager) {
+        this.executor = executor;
+        this.sqsClient = sqsClient;
+        this.config = config;
+        this.queueUrl = queueUrl;
+        this.queueAttributesManager = queueAttributesManager;
+    }
+
+    public void receiveMessage(ReceiveMessageCompletableFuture receiveMessageFuture) {
+        futures.add(receiveMessageFuture);
+        satisfyFuturesFromBuffer();
+        spawnMoreReceiveTasks();
+    }
+
+    public boolean isShutDown() {
+        return shutDown.get();
+    }
+
+    public void shutdown() {
+        if (this.shutDown.compareAndSet(false, true)) {
+            // Clear all finished tasks
+            while (!finishedTasks.isEmpty()) {
+                AsyncReceiveMessageBatch batch = finishedTasks.poll();
+                if (inflightReceiveMessageBatches.get() > 0) {
+                    inflightReceiveMessageBatches.decrementAndGet();
+                }
+                if (batch != null) {
+                    batch.clear();
+                }
+            }
+
+            // Clear futures
+            futures.forEach(future -> {
+                if (!future.responseCompletableFuture().isDone()) {
+                    future.setFailure(new CancellationException("Shutdown in progress"));
+                }
+            });
+            futures.clear();
+        }
+    }
+
+
+    private void spawnMoreReceiveTasks() {
+        if (shutDown.get()) {
+            return;
+        }
+
+        int desiredBatches = determineDesiredBatches();
+        if (finishedTasks.size() >= desiredBatches) {
+            return;
+        }
+
+        if (!finishedTasks.isEmpty() && (finishedTasks.size() + inflightReceiveMessageBatches.get()) >= desiredBatches) {
+            return;
+        }
+
+        queueAttributesManager.getVisibilityTimeout().thenAcceptAsync(visibilityTimeoutNanos -> {
```

Comment:
thenAcceptAsync by default executes on fork join pool, we should either do thenAccept or pass it to an executor

Reply:
good catch , passing executor now